core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,307 @@
+import software.amazon.awssdk.imds.MetadataResponse;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+@Immutable
+@ThreadSafe
+public final class DefaultEc2MetadataAsyncClient implements Ec2MetadataAsyncClient {
+
+    private static final Logger log = Logger.loggerFor(DefaultEc2MetadataClient.class);
+    private static final int DEFAULT_RETRY_THREAD_POOL_SIZE = 3;
+
+    private final Ec2MetadataRetryPolicy retryPolicy;
+    private final SdkAsyncHttpClient httpClient;
+    private final EndpointMode endpointMode;
+    private final URI endpoint;
+    private final RequestMarshaller requestMarshaller;
+    private final Duration tokenTtl;
+    private final ScheduledExecutorService asyncRetryScheduler;
+
+    private DefaultEc2MetadataAsyncClient(Ec2MetadataAsyncBuilder builder) {
+        this.retryPolicy = builder.retryPolicy != null ? builder.retryPolicy
+                                                       : Ec2MetadataRetryPolicy.builder().build();
+        this.httpClient = builder.httpClient != null ? builder.httpClient
+                                                     : AwsCrtAsyncHttpClient.create();
+        this.endpointMode = builder.endpointMode != null ? builder.endpointMode
+                                                         : DEFAULT_ENDPOINT_PROVIDER.resolveEndpointMode();
+        this.endpoint = builder.endpoint != null ? builder.endpoint
+                                                 : URI.create(DEFAULT_ENDPOINT_PROVIDER.resolveEndpoint(this.endpointMode));
+        this.requestMarshaller = new RequestMarshaller(this.endpoint);
+        this.tokenTtl = builder.tokenTtl != null ? builder.tokenTtl
+                                                 : Duration.ofSeconds(21600);
+        this.asyncRetryScheduler = builder.scheduledExecutorService != null
+                                   ? builder.scheduledExecutorService
+                                   : Executors.newScheduledThreadPool(DEFAULT_RETRY_THREAD_POOL_SIZE);
+    }
+
+    public static Ec2MetadataAsyncClient.Builder builder() {
+        return new DefaultEc2MetadataAsyncClient.Ec2MetadataAsyncBuilder();
+    }
+
+    @Override
+    public CompletableFuture<MetadataResponse> get(String path) {
+        return get(path, RetryPolicyContext.builder().retriesAttempted(0).build());
+    }
+
+    private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext retryPolicyContext) {
+        SdkHttpFullRequest baseTokenRequest = requestMarshaller.createTokenRequest(tokenTtl);
+
+        CompletableFuture<MetadataResponse> result = createAsyncRequest(baseTokenRequest)
+            .thenCompose(token -> {
+                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
+                return createAsyncRequest(baseMetadataRequest);
+            }).thenApply(MetadataResponse::create);
```

Comment:
Is it possible to rewrite this part of the code to make it a little easier to follow the request/response flow? I'm confused by the name `createAsyncRequest` (see other comment) and the token call is combined with the actual data call and it took me a while to realize what was happening. 

Maybe just splitting token and data futures like 
~~~
CompletableFuture<String> acquiredToken = callAsync(baseTokenRequest)
CompletableFuture<MetadataResponse> result =acquiredToken
            .thenCompose(token -> {
                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
                return createAsyncRequest(baseMetadataRequest);
            }).thenApply(MetadataResponse::create);
~~~